ckks/encoder.go

```diff
@@ -2,858 +2,1218 @@ package ckks
-	DecodeCoeffs(plaintext *rlwe.Plaintext) (res []float64)
-	DecodeCoeffsPublic(plaintext *rlwe.Plaintext, bound float64) (res []float64)
-
-	// Utility
-	Embed(values interface{}, logSlots int, scale rlwe.Scale, montgomery bool, polyOut interface{})
-	GetErrSTDCoeffDomain(valuesWant, valuesHave []complex128, scale rlwe.Scale) (std float64)
-	GetErrSTDSlotDomain(valuesWant, valuesHave []complex128, scale rlwe.Scale) (std float64)
-	ShallowCopy() Encoder
-	Parameters() Parameters
-}
+//	                                                    Z_Q[X]/(X^N+1)
+//		Coefficients: ---------------> Real^{N} ---------> Plaintext
+//	                                      |
+//	                                      |
+//		Slots: Complex^{N/2} -> iDFT -----┘
+type Encoder struct {
+	parameters Parameters
+
+	prec uint
 
-// encoder is a struct storing the necessary parameters to encode a slice of complex number on a Plaintext.
-type encoder struct {
-	params       Parameters
 	bigintCoeffs []*big.Int
 	qHalf        *big.Int
-	buff         *ring.Poly
+	buff         ring.Poly
 	m            int
 	rotGroup     []int
 
-	gaussianSampler *ring.GaussianSampler
-}
+	prng sampling.PRNG
 
-type encoderComplex128 struct {
-	encoder
-	values      []complex128
-	valuesFloat []float64
-	roots       []complex128
+	roots     interface{}
+	buffCmplx interface{}
 }
 
-// ShallowCopy creates a shallow copy of encoder in which all the read-only data-structures are
-// shared with the receiver and the temporary buffers are reallocated. The receiver and the returned
-// encoder can be used concurrently.
-func (ecd *encoder) ShallowCopy() *encoder {
+// NewEncoder creates a new Encoder from the target parameters.
+// Optional field `precision` can be given. If precision is empty
+// or <= 53, then float64 and complex128 types will be used to
+// perform the encoding. Else *big.Float and *bignum.Complex will be used.
+func NewEncoder(parameters Parameters, precision ...uint) (ecd *Encoder) {
 
-	prng, err := utils.NewPRNG()
-	if err != nil {
-		panic(err)
-	}
+	m := int(parameters.RingQ().NthRoot())
 
-	return &encoder{
-		params:          ecd.params,
-		bigintCoeffs:    make([]*big.Int, ecd.m>>1),
-		qHalf:           ring.NewUint(0),
-		buff:            ecd.params.RingQ().NewPoly(),
-		m:               ecd.m,
-		rotGroup:        ecd.rotGroup,
-		gaussianSampler: ring.NewGaussianSampler(prng, ecd.params.RingQ(), ecd.params.Sigma(), int(6*ecd.params.Sigma())),
-	}
-}
-
-// Parameters returns the parameters used by the encoder.
-func (ecd *encoder) Parameters() Parameters {
-	return ecd.params
-}
-
-func newEncoder(params Parameters) encoder {
-
-	m := int(params.RingQ().NthRoot())
-
-	rotGroup := make([]int, m>>1)
+	rotGroup := make([]int, m>>2)
 	fivePows := 1
 	for i := 0; i < m>>2; i++ {
 		rotGroup[i] = fivePows
 		fivePows *= int(GaloisGen)
 		fivePows &= (m - 1)
 	}
 
-	prng, err := utils.NewPRNG()
+	prng, err := sampling.NewPRNG()
+
+	// This error should never happen.
```

Comment:
Why because blake2b.NewXOF is unlikely to return an err?
By the way `NewPRNG` could return and error instead of panic for `rand.Read`.

Reply:
blake2b.NewXOF wil not return an error it will only return an error if the output length or seed are invalid and the implementation of NewPRNG guarantees that they are valid.

I've ensure that errors are returned by the packages `utils/sampling`.